SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriverRPC.cpp

```diff
@@ -0,0 +1,471 @@
+		{
+			It->DestroyValue_InContainer(Parms);
+		}
+	}
+
+	UFunction* Function;
+	uint8* Parms;
+};
+
+bool USpatialNetDriverRPC::ApplyRPC(Worker_EntityId EntityId, SpatialGDK::ReceivedRPC RPCData, const FRPCMetaData& MetaData)
+{
+	constexpr bool RPCConsumed = true;
+
+	FUnrealObjectRef ObjectRef(EntityId, RPCData.Offset);
+	TWeakObjectPtr<UObject> TargetObjectWeakPtr = NetDriver->PackageMap->GetObjectFromUnrealObjectRef(ObjectRef);
+	UObject* TargetObject = TargetObjectWeakPtr.Get();
+
+	const FClassInfo& ClassInfo = NetDriver->ClassInfoManager->GetOrCreateClassInfoByObject(TargetObjectWeakPtr.Get());
+	UFunction* Function = ClassInfo.RPCs[RPCData.Index];
+	if (Function == nullptr)
+	{
+		// TODO, log something
+		// return FRPCErrorInfo{ TargetObject, nullptr, ERPCResult::MissingFunctionInfo, ERPCQueueProcessResult::ContinueProcessing };
+		return RPCConsumed;
+	}
+
+	// NB : FMemory_Alloca is stack allocation, and cannot be done inside the RAII holder.
+	uint8* Parms = (uint8*)FMemory_Alloca(Function->ParmsSize);
+	RAIIParamsHolder ParamAlloc(Function, Parms);
+
+	TSet<FUnrealObjectRef> UnresolvedRefs;
+	{
+		// Scope for the SpatialNetBitReader lifecycle (only one should exist per thread).
+		TSet<FUnrealObjectRef> MappedRefs;
+
+		FSpatialNetBitReader PayloadReader(NetDriver->PackageMap, const_cast<uint8*>(RPCData.PayloadData.GetData()),
+										   RPCData.PayloadData.Num() * 8, MappedRefs, UnresolvedRefs);
+
+		TSharedPtr<FRepLayout> RepLayout = NetDriver->GetFunctionRepLayout(Function);
+		RepLayout_ReceivePropertiesForRPC(*RepLayout, PayloadReader, Parms);
+	}
+
+	const USpatialGDKSettings* SpatialSettings = GetDefault<USpatialGDKSettings>();
+
+	const float TimeQueued = (FPlatformTime::Cycles64() - MetaData.Timestamp) * FPlatformTime::GetSecondsPerCycle64();
+	const int32 UnresolvedRefCount = UnresolvedRefs.Num();
+
+	if (UnresolvedRefCount != 0 && TimeQueued < SpatialSettings->QueuedIncomingRPCWaitTime)
+	{
+		return !RPCConsumed;
+	}
+
+	if (UnresolvedRefCount > 0
+		// -> Use MetaData.ReceiverName to recover the RPC type.
+		/* && !SpatialSettings->ShouldRPCTypeAllowUnresolvedParameters(PendingRPCParams.Type)*/
+		&& (Function->SpatialFunctionFlags & SPATIALFUNC_AllowUnresolvedParameters) == 0)
+	{
+		const FString UnresolvedEntityIds = FString::JoinBy(UnresolvedRefs, TEXT(", "), [](const FUnrealObjectRef& Ref) {
+			return Ref.ToString();
+		});
+
+		UE_LOG(LogSpatialRPCService, Warning,
+			   TEXT("Executed RPC %s::%s with unresolved references (%s) after %.3f seconds of queueing. Owner name: %s"),
+			   *GetNameSafe(TargetObject), *GetNameSafe(Function), *UnresolvedEntityIds, TimeQueued,
+			   *GetNameSafe(TargetObject->GetOuter()));
+	}
+
+	// Get the RPC target Actor (was used for additional debug authority checking, but CanExtractRPC should handle that now)
+	// AActor* Actor = TargetObject->IsA<AActor>() ? Cast<AActor>(TargetObject) : TargetObject->GetTypedOuter<AActor>();
```

Comment:
why commented out?

Reply:
Some debug code leftover from a bit I copied. I wonder if it is still useful.

Reply:
I'd remove personally.